File: android/src/main/kotlin/io/supy/scanner/document/SupyDocumentScannerViewFactory.kt
package io.supy.scanner.document

import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import io.supy.scanner.barcode.ActivityHolder

/**
 * Factory for [SupyDocumentScannerView] PlatformViews.
 *
 * Registered against the view-type identifier `io.supy.scanner/v1/document_view`
 * — must match the Dart side in `SupyDocumentScannerView` (Dart).
 */
class SupyDocumentScannerViewFactory(
    private val messenger: BinaryMessenger,
    private val activityHolder: ActivityHolder,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        @Suppress("UNCHECKED_CAST")
        val creationParams = args as? Map<String, Any?>
        return SupyDocumentScannerView(
            context = context,
            viewId = viewId,
            creationParams = creationParams,
            messenger = messenger,
            activityHolder = activityHolder,
        )
    }

    companion object {
        const val VIEW_TYPE_ID = "io.supy.scanner/v1/document_view"
    }
}

File: ios/Classes/document/SupyDocumentScannerViewFactory.swift
import Flutter
import UIKit

/// Registers the embedded document scanner PlatformView.
///
/// View-type identifier and per-view channel scheme are defined in
/// `docs/ARCHITECTURE.md`. Mirrors `SupyBarcodeScannerViewFactory`.
final class SupyDocumentScannerViewFactory: NSObject, FlutterPlatformViewFactory {

  static let viewTypeId = "io.supy.scanner/v1/document_view"

  private let messenger: FlutterBinaryMessenger

  init(messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
    super.init()
  }

  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    let params = args as? [String: Any?]
    return SupyDocumentScannerView(
      frame: frame,
      viewId: viewId,
      creationParams: params,
      messenger: messenger
    )
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    return FlutterStandardMessageCodec.sharedInstance()
  }
}
